packages/medusa/src/strategies/batch-jobs/product/export.ts

```diff
@@ -0,0 +1,393 @@
+              .withTransaction(transactionManager)
+              .list(filterable_fields, {
+                ...list_config,
+                skip: offset,
+                take: limit,
+              } as FindProductConfig)
+          }
+
+          products.forEach((product: Product) => {
+            const lines = this.buildProductVariantLines(product)
+            lines.forEach((line) => writeStream.write(line))
+          })
+
+          advancementCount += products.length
+          offset += products.length
+          products.length = 0
+
+          batchJob = await this.batchJobService_
+            .withTransaction(transactionManager)
+            .update(batchJobId, {
+              context: {
+                ...batchJob.context,
+                file_key: fileKey,
+                list_config: {
+                  ...list_config,
+                  skip: offset,
+                },
+              },
+              result: {
+                count: productCount,
+                advancement_count: advancementCount,
+                progress: advancementCount / productCount,
+                err: undefined,
+              },
+            })
+
+          if (batchJob.status === BatchJobStatus.CANCELED) {
+            writeStream.end()
+
+            await this.fileService_
+              .withTransaction(transactionManager)
+              .delete({ key: fileKey })
+            return
+          }
+        }
+
+        writeStream.end()
+
+        return await promise
+      },
+      "REPEATABLE READ",
+      async (err) =>
+        this.handleProcessingErrors(batchJobId, err, {
+          offset,
+          count: productCount,
+          advancementCount,
+          progress: advancementCount / productCount,
+        })
+    )
+  }
+
+  private async buildHeader(): Promise<string> {
+    const transactionManager = this.transactionManager_ ?? this.manager_
+    const productRepo = transactionManager.getCustomRepository(
+      this.productRepository_
+    )
+
+    const [{ maxOptionsCount, maxImagesCount, maxMoneyAmountCount }] =
+      await productRepo.query(`
+        WITH "optionsShape" AS (
+            SELECT count(*) as "maxOptionsCount"
+            from product_option
+            group by product_id
+            order by "maxOptionsCount" DESC
+            LIMIT 1
+        ), "imagesShape" AS (
+            SELECT count(*) as "maxImagesCount"
+            from product_images
+            group by product_id
+            order by "maxImagesCount" DESC
+            LIMIT 1
+        ), "moneyAmountShape" AS (
+            SELECT count(*) as "maxMoneyAmountCount"
+            from money_amount
+            group by variant_id
+            order by "maxMoneyAmountCount" DESC
+            LIMIT 1
+        )
+        SELECT 
+           "maxOptionsCount",
+           "maxImagesCount",
+           "maxMoneyAmountCount"
+        FROM 
+           "optionsShape",
+           "imagesShape",
+           "moneyAmountShape";
+      `)
```

Comment:
It is very specific and related to the product export so I though it would not have to be elsewhere. I could use the manager instead of the productRepo in fact and that way it remove that injection. Wdyt?

Reply:
My thinking is that we should NOT allow direct access to the database except for through services (in turn services should only be allowed to read from corresponding repos, but that is unrelated here). 

This will ensure that we only have one place to change things that perform operations on the db, which is really nice for the readability and structure of the code base but it also has the added benefit of giving us a good amount of optionality in future modularization.

Imagine - hypothetically speaking - that we want to allow customers to use 3rd party tools to store the product data, e.g. in a PIM like Akeneo. If we have all data access through services the modifications needed to accomplish this would be much more lightweight.

Reply:
Yes I am completely  agree, I did not contradict t what you were saying since I understood, Just explained why I did that 😂

Reply:
I will move it :)